cruft/_commands/utils/generate.py

```diff
@@ -0,0 +1,116 @@
+from typing import Optional, Set
+
+from cookiecutter.generate import generate_files
+from git import Repo
+
+from .cookiecutter import CookiecutterContext, generate_cookiecutter_context
+from .cruft import CruftState
+
+try:
+    import toml  # type: ignore
+except ImportError:  # pragma: no cover
+    toml = None  # type: ignore
+
+
+def cookiecutter_template(
+    output_dir: Path,
+    repo: Repo,
+    cruft_state: CruftState,
+    project_dir: Path = Path("."),
+    cookiecutter_input: bool = False,
+    checkout: Optional[str] = None,
+) -> CookiecutterContext:
+    """Generate a clean cookiecutter template in output_dir."""
+    pyproject_file = project_dir / "pyproject.toml"
+    commit = checkout or repo.remotes.origin.refs["HEAD"]
+
+    repo.head.reset(commit=commit, working_tree=True)
+
+    output_dir.mkdir(parents=True, exist_ok=True)
+    context = _generate_output(cruft_state, Path(repo.working_dir), cookiecutter_input, output_dir)
+
+    # Get all paths that we are supposed to skip before generating the diff and applying updates
+    skip_paths = _get_skip_paths(cruft_state, pyproject_file)
+    # We also get the list of paths that were deleted from the project
+    # directory but were present in the template that the project is linked against
+    # This is to avoid introducing changes that won't apply cleanly to the current project.
+    deleted_paths = _get_deleted_files(output_dir, project_dir)
+    # We now remove skipped and deleted paths from the project
+    _remove_paths(output_dir, skip_paths | deleted_paths)
+
+    return context
+
+
+#####################################
+# Generating clean outputs for diff #
+#####################################
+
+
+def _generate_output(
+    cruft_state: CruftState, project_dir: Path, cookiecutter_input: bool, output_dir: Path
+) -> CookiecutterContext:
+    inner_dir = project_dir / (cruft_state.get("directory") or "")
+
+    new_context = generate_cookiecutter_context(
+        cruft_state["template"],
+        inner_dir,
+        extra_context=cruft_state["context"]["cookiecutter"],
+        no_input=not cookiecutter_input,
+    )
+
+    # This generates the cookiecutter template.
+    # Unfortunately, cookiecutter doesn't let us output the template in an
+    # arbitrary directory. It insists on creating the initial project directory.
+    # Therefore we have to move the directory content to the expected output_dir.
+    # See https://github.com/cookiecutter/cookiecutter/pull/907
+    with TemporaryDirectory() as tmpdir_:
+        tmpdir = Path(tmpdir_)
+
+        # Kindly ask cookiecutter to generate the template
+        template_dir = generate_files(
+            repo_dir=inner_dir, context=new_context, overwrite_if_exists=True, output_dir=tmpdir
+        )
+        template_dir = Path(template_dir)
+
+        # Move the template content to the output directory
+        for name in os.listdir(template_dir):
+            move(str(template_dir / name), str(output_dir))
```

Comment:
IIUC, this might fail if the generated folder has a subfolder with the same name?

For eg, what would happen if template_dir looked like -

```
test
└── test
    └── asd
--
```

Where output_dir is on the same level as the top level `test` dir.

Reply:
I may be missing something but I don't see why it would. In you exemple:
- the loop would make one iteration because `os.listdir` is shallow and `template_dir` contains one entry (the inner `test` directory)
- `template_dir` would equal, say, `/tmpdir/xxx/test`
- `name` would equal `test`
- `output_dir` would equal, say, `/tmpdir/yyy`

So at the end of the day, one call would be issued to the move function and this call would be equivalent to `shutil.move("/tmpdir/xxx/test/test", "/tmpdir/yyy")` which, as far as I understand, is what we want. The inner `test` directory would be moved to the `output_dir`.

Does that make sense? Am I missing something?

Reply:
Oh, I think I see what you meant. The `_générateur_template` function expects the `output_dir` to exist which is gross. I'll move the mkdir call in this function as soon as I can reach a computer.

Reply:
Does [this](https://github.com/cruft/cruft/pull/59/commits/d772a9d02f7da16e4d3b2cd8662b222ff0054c73) address your remark?